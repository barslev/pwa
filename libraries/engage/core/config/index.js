export { default as subscriptions } from './config.subscriptions';
export { default as reducers } from './config.reducers';

export * from './config.actions';
export * from './config.selectors';
export * from './config.streams';
