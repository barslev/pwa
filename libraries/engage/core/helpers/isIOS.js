// @flow
/**
 * Checks if current theme is an iOS theme.
 * @returns {boolean}
 */
export function isIOS() {
  if (!process.env.THEME) {
    return false;
  }

  return process.env.THEME.includes('ios');
}
