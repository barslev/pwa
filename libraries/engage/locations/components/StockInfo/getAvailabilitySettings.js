import { availabilityTypes } from '@shopgate/engage/product';

/**
 * Takes the location stock info settings and finds the matching settings for the given inventory.
 * @param {Object} settings The settings find a match in.
 * @param {number|null} visibleInventory The visible inventory defines, which settings to use.
 * @param {number|null} inventoryBlind The blind inventory defines as well, which settings to use.
 * @returns {Object}
 */
export default (settings, visibleInventory, inventoryBlind) => {
  let matchingTypes;
  if (visibleInventory === null) {
    // Since no availability is given, check if any types are configured to be filtered by that.
    matchingTypes = availabilityTypes.filter(type => (
      !!type.includeNoRecordStores
    ));

    if (matchingTypes.length > 0) {
      return settings[matchingTypes[0]];
    }
  }

  // Filter by inventory blind and visible inventory (must match both).
  matchingTypes = availabilityTypes
    .filter(type => (
      // When inventory blind is set in the current availability setting, then this should also
      // account for the inventory blind from the given store. Ignore inventory blind otherwise.
      !settings[type].includeInventoryBlind ||
      (settings[type].includeInventoryBlind && !!inventoryBlind)
    ))
    .filter(type => (
      // Don't filter, when inventoryBlind filter is already active
      settings[type].includeInventoryBlind || (
        // Also, inventory must be truthy or 0, or otherwise there is nothing to search for.
        visibleInventory >= settings[type].visibleInventoryFrom &&
        (settings[type].visibleInventoryTo === null ||
          visibleInventory <= settings[type].visibleInventoryTo)
      )
    ));

  if (matchingTypes.length > 0) {
    return settings[matchingTypes[0]];
  }

  return {};
};
