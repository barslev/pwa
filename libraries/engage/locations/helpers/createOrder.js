import { getProductDataById } from '@shopgate/pwa-common-commerce/product/selectors/product';
import { makeGetUserLocation } from '../selectors';

/**
 * Creates the address sequence.
 * @param {Object} formValues The reserve form values.
 * @returns {Array}
 */
function createAddressSequence(formValues) {
  return [
    {
      type: 'billing',
      firstName: formValues.firstName,
      lastName: formValues.lastName,
      phone: formValues.cellPhone,
      emailAddress: formValues.email,
    },
    {
      type: 'pickup',
      firstName: formValues.firstName2,
      lastName: formValues.lastName2,
      phone: formValues.cellPhone2,
      emailAddress: formValues.email2,
    },
  ];
}

/**
 * Retrieves the featured image for a product. If the product doesn't have a featured image
 * and is a child product, the featured image of the base product is retrieved.
 * @param {Object} product The current product.
 * @param {FUnction} getState The redux getState function.
 * @returns {string}
 */
function getProductImage(product, getState) {
  let image = product.featuredImageUrl;

  if (!image && product.baseProductId) {
    const baseProduct = getProductDataById(getState(), { productId: product.id });

    if (baseProduct) {
      image = baseProduct.featuredImageUrl;
    }
  }

  return image;
}

/**
 * Creates the order line items.
 * @param {Object} product The current product.
 * @param {Object} location The selected user location.
 * @param {Function} getState The redux getState function.
 * @returns {Array}
 */
function createLineItems(product, location, getState) {
  return [
    {
      code: product.id,
      quantity: 1, // TODO: can be different from cart, once we implement reserve to cart.
      fulfillmentMethod: 'ROPIS', // TODO: Change based on the selected method.
      fulfillmentLocationCode: location.code,
      price: product.price.unitPrice,
      shipToAddressSequenceIndex: 0,
      currencyCode: product.price.currency,
      product: {
        code: product.id,
        name: product.name,
        image: getProductImage(product, getState),
        price: product.price.unitPrice,
        currencyCode: product.price.currency,
      },
    },
  ];
}

/**
 * Creates a fulfillment order out of the reserve form values and the current product.
 * @param {Object} formValues The reserve form values.
 * @param {Object} product The current product.
 * @param {Function} getState The redux getState function.
 * @returns {Object}
 */
function createOrder(formValues, product, getState) {
  const location = makeGetUserLocation()(getState());

  // create order
  const order = {
    // localeCode: process.env.LOCALE_FILE.replace('-', '_'),
    localeCode: 'de_DE',
    currencyCode: product.price.currency,
    addressSequences: createAddressSequence(formValues),
    primaryBillToAddressSequenceIndex: 0,
    lineItems: createLineItems(product, location, getState),
    subTotal: product.price.unitPrice,
    total: product.price.unitPrice,
  };

  return order;
}

export default createOrder;
