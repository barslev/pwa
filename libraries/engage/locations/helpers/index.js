export { default as createOrder } from './createOrder';
export { default as getAvailabilitySettings } from './getAvailabilitySettings';
