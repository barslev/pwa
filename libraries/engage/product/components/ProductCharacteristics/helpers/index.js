import isMatch from 'lodash/isMatch';
import appConfig from '@shopgate/pwa-common/helpers/config';

const { variantSelectionMode, product: { variantPreselect } = {} } = appConfig;
const preselectVariant = variantPreselect || parseInt(variantSelectionMode, 10) === 1;

/**
 * Returns the index of a particular characteristic from a set of characteristics.
 * @param {Array} characteristics The characteristics of a product.
 * @param {number} characteristicId The id of the characteristic to find.
 * @return {number}
 */
export function findSelectionIndex(characteristics, characteristicId) {
  return characteristics.findIndex(char => char.id === characteristicId);
}

/**
 * Returns true if the characteristic at the given index is the next characteristic to be selected.
 * @param {Array} characteristics The characteristics of a product.
 * @param {number} index The index of the characteristic to check.
 * @return {boolean}
 */
export function isCharacteristicEnabled(characteristics, index) {
  if (index === 0) {
    return true;
  }

  return !!Object.values(characteristics)[index - 1];
}

/**
 * Returns the selected value by characteristic ID.
 * @param {string} charId The characteristic ID.
 * @param {Object} characteristics The already selected characteristics.
 * @return {string|null}
 */
export function getSelectedValue(charId, characteristics) {
  return characteristics[charId] ? characteristics[charId] : null;
}

/**
 * Prepares the new state after a selection has been made.
 * @param {string} id The selection ID
 * @param {string} value The selection value.
 * @param {Object} selections The selections stored in the state.
 * @param {Array} characteristics The characteristics of a product.
 * @param {Array} products All available products.
 * @return {Object}
 */
export function prepareState(id, value, selections, characteristics, products) {
  const updateValid = !!characteristics.find(({ id: cId, values }) => {
    if (cId !== id) {
      return false;
    }
    return !!values.find(({ id: vId }) => vId === value);
  });

  if (!updateValid) {
    // Input parameters are invalid.
    return selections;
  }

  // Merge the just changed value into the previous selection.
  const currentSelection = {
    ...selections,
    [id]: value,
  };

  /**
   * Determine if there are products for the current user selection.
   */
  const matches = products.filter(product => (
    isMatch(product.characteristics, currentSelection)
  ));

  if (matches.length === 1) {
    // Product found for the current selection.
    return { ...matches[0].characteristics };
  }

  if (!selections[id]) {
    // Changed value wasn't set before.
    return currentSelection;
  }

  // Find the current index.
  const index = findSelectionIndex(characteristics, id);

  /**
   * When the given index is the same as the complete set
   * of characteristics then there is nothing to reset.
   */
  if (index === (characteristics.length - 1)) {
    return currentSelection;
  }

  // Find the selections after the given index.
  const after = characteristics.slice(index + 1);

  // Delete the found selections.
  after.forEach((item) => {
    delete currentSelection[item.id];
  });

  return currentSelection;
}

/**
 * Preselect characteristics for variant
 * or pre-select the first available product
 * @param {string} [variantId=null] The selected variant
 * @param {{products: Object[], characteristics: Object[]}} [variants=null] All possible variants.
 * @return {Object}
 */
export function selectCharacteristics({ variantId, variants = {} }) {
  if (!variants || !variants.products || !variants.products.length) {
    return {};
  }

  if (variantId) {
    const variant = variants.products.find(product => product.id === variantId) || {};
    return { ...variant.characteristics };
  }

  // If product has only 1 variant preselect no matter if "preselect" is chosen or not.
  if (variants.products.length === 1) {
    return { ...variants.products[0].characteristics };
  }

  // Pre-selection is off
  if (!preselectVariant) {
    return {};
  }

  // Find the first selectable product by characteristics
  return variants.characteristics.reduce((acc, char) => {
    // Find the first char value with selectable products
    const firstVal = char.values.find((val) => {
      // eslint-disable-next-line extra-rules/no-single-line-objects
      const source = { ...acc, [char.id]: val.id };
      return variants.products.filter(p => (
        isMatch(p.characteristics, source)
      )).length > 0;
    });

    if (!firstVal) {
      return acc;
    }
    // eslint-disable-next-line extra-rules/no-single-line-objects
    return { ...acc, [char.id]: firstVal.id };
  }, {});
}
