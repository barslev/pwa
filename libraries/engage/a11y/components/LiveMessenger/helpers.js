import { UIEvents } from '../../../core';
import { EVENT_LIVE_MESSAGE, LIVE_MESSAGE_TYPE_POLITE } from './constants';

/**
 * Broadcasts a live message to a LiveMessenger component.
 * @param {string} message The message.
 * @param {Object} options Additional options.
 * @param {string} options.type Type of the message
 * @param {Object} options.params Message params for an i18n placeholder message.
 * @param {string} options.id Id of the LiveMessenger component which sets the message.
 * @param {boolean} options.force If set to TRUE the LiveMessenger component will read the text,
 * even if the same was broadcasted before.
 */
export function broadcastLiveMessage(message, options = {}) {
  const defaults = {
    type: LIVE_MESSAGE_TYPE_POLITE,
    params: null,
    id: null,
  };

  if (!message) {
    return;
  }

  const params = {
    ...defaults,
    ...options,
  };

  UIEvents.emit(EVENT_LIVE_MESSAGE, message, params);

  setTimeout(() => {
    // Clear the live area after a short time, so that the screen reader can't focus the element.
    UIEvents.emit(EVENT_LIVE_MESSAGE, '', params);
  }, 100);
}
