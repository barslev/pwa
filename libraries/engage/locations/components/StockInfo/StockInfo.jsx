import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'glamor';
import classNames from 'classnames';
import defaultsDeep from 'lodash/defaultsDeep';
import { getThemeSettings } from '../../../core';
import { SurroundPortals } from '../../../components';
import { PRODUCT_LOCATION_STOCK_INFO } from '../../constants';
import { getAvailabilitySettings } from '../../helpers';
import defaultSettings from './StockInfo.defaultSettings';
import Inventory from './StockInfoInventory';
import StoreName from './StockInfoStoreName';

/**
 * Renders visible stock information based on the given location.
 * @param {Object} props The component props.
 * @return {JSX}
 */
function StockInfo({ location, className, showStoreName }) {
  const { locationStockInfo } = getThemeSettings('product') || {};
  const settings = defaultsDeep(locationStockInfo, defaultSettings);

  const { availabilityText = '', availabilityTextColor = 'inherit' } =
    getAvailabilitySettings(settings, location);

  const defaultClassName = css({
    color: availabilityTextColor,
    fontSize: '0.875rem',
    margin: 0,
  });

  let displayCapitalized = location.productInventory && location.productInventory.visible === null;
  displayCapitalized = displayCapitalized || !availabilityText;

  return (
    <SurroundPortals
      portalName={PRODUCT_LOCATION_STOCK_INFO}
      portalProps={{
        location,
        className,
        availabilityText,
        availabilityTextColor,
      }}
    >
      <span className={classNames(`${defaultClassName}`, `${className}`)}>
        <Inventory
          availabilityText={availabilityText}
          location={location}
          maxNumberVisible={settings.maxNumberOfVisibleInventory}
          aboveMaxExtension={settings.aboveMaxExtension}
        />
        <StoreName
          name={showStoreName ? location.name : ''}
          displayCapitalized={displayCapitalized}
        />
      </span>
    </SurroundPortals>
  );
}

StockInfo.propTypes = {
  location: PropTypes.shape({
    name: PropTypes.string,
    productInventory: PropTypes.shape({
      visible: PropTypes.number,
    }),
  }).isRequired,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape(),
  ]),
  showStoreName: PropTypes.bool,
};

StockInfo.defaultProps = {
  className: '',
  showStoreName: true,
};

export default StockInfo;
