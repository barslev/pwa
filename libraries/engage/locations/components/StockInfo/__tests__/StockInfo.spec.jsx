import React from 'react';
import { mount } from 'enzyme';
import { getThemeSettings } from '../../../../core';
import { getAvailabilitySettings } from '../../../helpers';
import StockInfo from '../StockInfo';

jest.mock('../../../../core', () => ({
  ...jest.requireActual('../../../../core'),
  getThemeSettings: jest.fn(),
}));
jest.mock('../../../helpers', () => ({
  getAvailabilitySettings: jest.fn(),
}));

describe('<StockInfo />', () => {
  it('should not render if no location data is passed', () => {
    getThemeSettings.mockReturnValueOnce({});
    getAvailabilitySettings.mockReturnValueOnce({
      availabilityText: 'demo_availability_text',
      availabilityTextColor: 'green',
    });
    const wrapper = mount(<StockInfo
      location={{
        name: '',
        visibleInventory: null,
      }}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').length).toBe(0);
  });

  it('should render with the visible inventory and with the store name', () => {
    getThemeSettings.mockReturnValueOnce({});
    getAvailabilitySettings.mockReturnValueOnce({
      availabilityText: 'demo_availability_text',
      availabilityTextColor: 'green',
    });
    const wrapper = mount(<StockInfo
      location={{
        name: 'Test Store',
        visibleInventory: 15,
      }}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('[string="demo_availability_text"]').length).toBe(1);
    expect(wrapper.find('[string="product.location_stock_info.pick_up_at"]').length).toBe(1);
  });

  it('should render without store name', () => {
    getThemeSettings.mockReturnValueOnce({});
    getAvailabilitySettings.mockReturnValueOnce({
      availabilityText: 'demo_availability_text',
      availabilityTextColor: 'green',
    });
    const wrapper = mount(<StockInfo
      location={{
        name: 'Test Store',
        visibleInventory: 15,
      }}
      showStoreName={false}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('[string="product.location_stock_info.pick_up_at"]').length).toBe(0);
  });

  it('should render without the visible inventory', () => {
    getThemeSettings.mockReturnValueOnce({});
    getAvailabilitySettings.mockReturnValueOnce({
      availabilityText: 'demo_availability_text',
      availabilityTextColor: 'green',
    });
    const wrapper = mount(<StockInfo
      location={{
        name: 'Test Store',
        visibleInventory: null,
      }}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('[string="demo_availability_text"]').length).toBe(0);
  });
});
