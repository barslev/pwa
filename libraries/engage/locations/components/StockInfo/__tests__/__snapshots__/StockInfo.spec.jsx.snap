// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StockInfo /> should render with the visible inventory and with the store name 1`] = `
<StockInfo
  className=""
  location={
    Object {
      "name": "Test Store",
    }
  }
  showStoreName={true}
>
  <SurroundPortals
    portalName="product.location-stock-info"
    portalProps={
      Object {
        "availabilityText": "demo_availability_text",
        "availabilityTextColor": "green",
        "className": "",
        "location": Object {
          "name": "Test Store",
        },
      }
    }
  >
    <Portal
      name="product.location-stock-info.before"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
    <Portal
      name="product.location-stock-info"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    >
      <span
        className="css-d3hu7n"
      >
        <Inventory
          aboveMaxExtension="++"
          availabilityText="demo_availability_text"
          location={
            Object {
              "name": "Test Store",
            }
          }
          maxNumberVisible={100}
        />
        <StoreName
          displayCapitalized={false}
          name="Test Store"
        >
          <Translate
            className={null}
            params={
              Object {
                "storeName": "Test Store",
              }
            }
            role={null}
            string="product.location_stock_info.pick_up_at"
            transform={null}
          >
            <span
              className={null}
              role={null}
            >
              product.location_stock_info.pick_up_at
            </span>
          </Translate>
        </StoreName>
      </span>
    </Portal>
    <Portal
      name="product.location-stock-info.after"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
  </SurroundPortals>
</StockInfo>
`;

exports[`<StockInfo /> should render without store name 1`] = `
<StockInfo
  className=""
  location={
    Object {
      "name": "Test Store",
    }
  }
  showStoreName={false}
>
  <SurroundPortals
    portalName="product.location-stock-info"
    portalProps={
      Object {
        "availabilityText": "demo_availability_text",
        "availabilityTextColor": "green",
        "className": "",
        "location": Object {
          "name": "Test Store",
        },
      }
    }
  >
    <Portal
      name="product.location-stock-info.before"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
    <Portal
      name="product.location-stock-info"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    >
      <span
        className="css-d3hu7n"
      >
        <Inventory
          aboveMaxExtension="++"
          availabilityText="demo_availability_text"
          location={
            Object {
              "name": "Test Store",
            }
          }
          maxNumberVisible={100}
        />
        <StoreName
          displayCapitalized={false}
          name=""
        />
      </span>
    </Portal>
    <Portal
      name="product.location-stock-info.after"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
  </SurroundPortals>
</StockInfo>
`;

exports[`<StockInfo /> should render without the visible inventory 1`] = `
<StockInfo
  className=""
  location={
    Object {
      "name": "Test Store",
    }
  }
  showStoreName={true}
>
  <SurroundPortals
    portalName="product.location-stock-info"
    portalProps={
      Object {
        "availabilityText": "demo_availability_text",
        "availabilityTextColor": "green",
        "className": "",
        "location": Object {
          "name": "Test Store",
        },
      }
    }
  >
    <Portal
      name="product.location-stock-info.before"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
    <Portal
      name="product.location-stock-info"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    >
      <span
        className="css-d3hu7n"
      >
        <Inventory
          aboveMaxExtension="++"
          availabilityText="demo_availability_text"
          location={
            Object {
              "name": "Test Store",
            }
          }
          maxNumberVisible={100}
        />
        <StoreName
          displayCapitalized={false}
          name="Test Store"
        >
          <Translate
            className={null}
            params={
              Object {
                "storeName": "Test Store",
              }
            }
            role={null}
            string="product.location_stock_info.pick_up_at"
            transform={null}
          >
            <span
              className={null}
              role={null}
            >
              product.location_stock_info.pick_up_at
            </span>
          </Translate>
        </StoreName>
      </span>
    </Portal>
    <Portal
      name="product.location-stock-info.after"
      props={
        Object {
          "availabilityText": "demo_availability_text",
          "availabilityTextColor": "green",
          "className": "",
          "location": Object {
            "name": "Test Store",
          },
        }
      }
    />
  </SurroundPortals>
</StockInfo>
`;
